### PR TITLE
CODAP-1313: optimize marquee-selection performance

### PR DESCRIPTION
**Pull request by kswenson**
## Summary

Optimizes marquee point-selection in the graph (and the linked case table's
response). Profiled with the built-in `prf` profiler, the per-marquee-move
selection cost drops from **~23 ms/move to ~1.75 ms/move (~13×)** on large data,
with the win scaling as dataset size grows.

The work, in layers:
- **Coalesce per frame** — graph point-selection refresh and case-table
  row-selection sync are batched to one update per animation frame instead of
  running synchronously on every incremental `selectCases`.
- **Delta restyle** — restyle only the points whose selection changed, not all N.
- **Renderer-effect gate** — the renderer-availability effect refreshes only when
  the renderer actually changes (it was doing a full restyle on every re-render,
  masking the above).
- **Volatile selection fast path** — during a marquee, update the already-volatile
  selection via the raw `selectCases` action and notify plugins directly from the
  known delta, skipping the per-move `applyModelChange` + `selectCasesNotification`
  wrapper (history boundary + O(selection) before/after snapshot). Selection isn't
  undoable and is serialized at save, so no drag-end commit is needed; per-move
  plugin notifications are preserved (now O(delta)).
- **Eliminate O(N) selection-computed recomputes** — the dominant remaining cost was
  `DataConfiguration.selection`/`unselectedCases` (O(N)) recomputing every move
  because always-mounted observers read them. Gated the graph hide/show menu (via a
  new menu open-state context), the measures-for-selection reaction, and the
  connecting-lines autorun to observe selection only when their feature is active.

### Code-review fixes (high-effort review on this branch)
- Point-selection restyle made robust to non-delta selection changes (parent /
  legend / plugin selections and snapshot/moment restores now restyle correctly).
- Case table scrolls to marquee-*selected* cases (not deselected); fixed a
  wrong closest-row distance calc.
- Plugins notified only about cases whose selection actually changes.

This is V3-only; the renderer choice (WebGL vs canvas) doesn't affect the
selection path (verified). A separate `pixiRender` per-frame cost (z-sorting raised
sprites on large selections) is noted as a follow-up, distinct from selection.

## Test plan
- [x] `npm run build:tsc` and `npm run lint` clean
- [x] Jest: graph / data-display / case-table / data-model suites green (~1187 tests)
- [x] Manual: marquee-select on a large scatter plot + linked case table — points
      recolor (delta), table scrolls to selected cases, no jank; legend shift-click
      and individual selection still correct; verified under both WebGL and canvas
- [ ] CI (lint, build, Cypress) + full regression

Fixes CODAP-1313.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: v3/src/models/data/data-set-utils.ts [hunk 1]
```diff
@@ -182,6 +183,28 @@ export function selectAndDeselectCases(addCaseIds: string[], removeCaseIds: stri
...
+) {
+  if (!data) return
+  // Notify only about cases whose selection state actually changes (mirrors selectCasesNotification's
+  // before/after diff), computed before mutating — an oscillating/overlapping marquee can re-include
+  // already-selected (or already-deselected) ids, which should not be reported to plugins.
+  const addedCaseIds = addCaseIds.filter(id => !data.isCaseSelected(id))
+  const removedCaseIds = removeCaseIds.filter(id => data.isCaseSelected(id))
+  addCaseIds.length && data.selectCases(addCaseIds)
+  removeCaseIds.length && data.selectCases(removeCaseIds, false)
+  const note = selectCasesNotificationForDelta(data, addedCaseIds, removedCaseIds)
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/models/data/data-set-utils.ts:198]: Good catch — fixed in 3ced2370d. The mutation now uses the filtered addedCaseIds/removedCaseIds and early-returns when neither actually changes, so re-adding an already-selected case no longer bumps selectionChanges or triggers refreshes.

File: v3/src/components/data-display/renderer/pixi-point-renderer.ts [hunk 2]
```diff
@@ -613,7 +614,9 @@ export class PixiPointRenderer extends PointRendererBase {
       this.ticker.stop()
       this.cleanupUnusedTextures()
     }
-    this.renderer.render(this.stage)
+    prf.measure("Graph.pixiRender", () => {
+      this.renderer?.render(this.stage)
+    })
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/data-display/renderer/pixi-point-renderer.ts]: Thanks — addressed in 519d79592, at the profiler level rather than this call site: prf.measure() and prf.end() now early-return when profiling is off, so this wrapper (and every other prf.measure site, including the ones this PR added on hot paths) has no per-frame overhead when idle, and the wrapper stays clean.

File: v3/src/components/data-display/renderer/canvas-point-renderer.ts [hunk 3]
```diff
@@ -444,7 +445,9 @@ export class CanvasPointRenderer extends PointRendererBase {
 
     // Render if needed
     if (this.needsRedraw) {
-      this.drawAllPoints()
+      prf.measure("Graph.canvasRender", () => {
+        this.drawAllPoints()
+      })
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/data-display/renderer/canvas-point-renderer.ts]: Same fix as the pixi renderer — 519d79592 makes prf.measure/end zero-overhead when not profiling, so this drawAllPoints() wrapper has no per-frame cost in normal runs.

**Review by bfinzer — APPROVED** [review]
👍🏻 LGTM
My only question is whether it's a good idea to leave all the `prf.measure` in. I don't know if they affect performance, but they certainly clutter up the code a bit.

**Comment by kswenson** [comment]
Good point — trimmed the fine-grained breakdowns that were specific to this marquee dig (the `dragMarquee[setRect]/[diff]/[selectCases]` and `setPointCoordinates[default]/[transition]` sub-measures, plus the `[delta]` label variants). Left the coarse per-operation measures in place for hot-spot identification, since `prf.measure` is now zero-overhead when profiling is off and matches the existing convention elsewhere in the codebase. We can re-add the finer breakdowns next time we profile marquee selection specifically. Thanks!

**Comment by kswenson** [comment]
**Follow-up (commit `5aeeced8f`): collapsed the marquee to a single selection path.**

While reviewing the perf changes against V2 (`origin/master`) for parity, I found that the two V3 marquee selection paths diverged in their plugin notifications on hierarchical data:

- The **performance path** (`selectCasesNotificationForDelta`) reports only the childmost plotted cases in the `selectCases` delta.
- The old **non-performance path** (`selectCasesNotification`) scanned all cases and additionally reported **parent collection cases** whose children had all become selected.

V2's marquee (`graph_view.js` → `doSelectCases` → `notification_manager.contextDataDidChange`) reports **childmost cases only** — it deliberately does *not* auto-select a parent from its children (see the explicit comment in V2's `doSelectCases`). So the performance path is the V2-faithful one; the non-perf path was the deviation.

Since the non-perf branch was also unreachable in production (performance mode is always enabled during a drag) and — unlike the scatter-plot point-drag fast path — carries no riskier rendering model to fall back to, I removed it: deleted the dead `selectAndDeselectCases` wrapper, renamed `selectAndDeselectCasesInteractive` → `selectAndDeselectCases`, and call it unconditionally. Behavior-preserving in production; `tsc`/lint/affected unit tests green.

**QA note:** the highest-value manual check is a hierarchical-data marquee with a selection-aware plugin — confirm V3's streamed `selectCases` notifications now match V2's (childmost-only).

**Comment by kswenson** [comment]
**Merge-readiness: risk analysis + manual test plan**

This is a performance PR, so the failure mode to guard against is "does too little" — a coalesced/delta/flag-gated path that skips an update some edge case needs. Here's where that risk lives and what QA should confirm.

**What changed (risk-bearing parts):**
- Marquee selection now uses a single volatile fast path (`selectAndDeselectCases`) — raw `selectCases` + a direct, delta-built plugin notification, no `applyModelChange`. This is the **default** for every drag.
- Point restyle is now a per-frame **delta** (only changed cases) with a full-restyle fallback; RDG row-selection sync and graph point-selection refresh are **rAF-coalesced**.
- Several reactions now observe the O(N) selection **only when their feature is enabled** (measures-for-selection, connecting lines, hide/show menu counts).

**V2 parity (the one intentional behavior change):** on hierarchical data the marquee's streamed `selectCases` notification is now **childmost-only**, which **matches V2** (`graph_view.js` → `doSelectCases` → `notification_manager`, which deliberately does not auto-select parents from children). The previous non-perf path's parent-inclusion was the V3 deviation and has been removed.

**Residual risk to watch — the renderer-availability effect (`use-plot.ts`).** Its early-return (only refresh on renderer change) is the right perf call, but it removed a previously-unconditional **post-commit** refresh that other code was incidentally relying on. The Squares of Residuals regression surfaced by the rebase was one consequence (now fixed with a dedicated effect). Any other graph visual that leaned on that incidental refresh — rather than on a dedicated data/selection/layout reaction — could be similarly affected. This is the main reason the **full Cypress regression is the key merge gate**, and why manual QA should watch broadly for any graph element that fails to update after a selection / data / layout / renderer change.

**Already verified:** full Jest suite green (3284 passing); `tsc` + lint clean; the squares-of-residuals regression is fixed and the `bivariate-adornments` spec passes 8/8 locally.

**Manual test plan (highest-value first):**
1. **Marquee correctness, large flat data:** rapid + oscillating drags; confirm no "stuck" highlighted points after the rectangle moves off them, and the final selection exactly matches the rectangle. Repeat on dot/bar/histogram/binned plots and a **two-y-attribute scatter** (the multi-plot delta path).
2. **Hierarchical marquee + a selection-aware plugin:** confirm V3's streamed `selectCases` notifications now match V2's (childmost-only). This is the one behavioral change — if your target plugins don't consume parent-case selection granularity, it's effectively cosmetic.
3. **Cross-tile sync:** graph↔case-table selection highlight/scroll; confirm the table scrolls toward *selected* rows and **not** on deselect.
4. **Feature-flag gates:** make a selection with Measures-for-Selection / Connecting Lines **off**, then turn each **on** — confirm they reflect the existing selection.
5. **Adornment / stale-visual regression guard:** Squares of Residuals (movable line, LSRL, plotted function) render on toggle-on (the exact failure the rebase surfaced); more generally, watch any adornment or point styling that should update after selection/data/layout changes.
6. **Persistence:** marquee-select → save → reload restores the selection.

**Merge gate:** green full Cypress regression on CI, plus a human eye on checks #2 and #5. With those, I'd consider this safe to merge.

**Comment by kswenson** [comment]
**Follow-up opportunities (non-blocking, for a later cleanup pass)**

None of these block this PR; capturing them so they aren't lost.

- **Generalize the "observe selection only when feature X is on" gating.** It's now hand-applied per consumer (measures-for-selection, connecting lines, and the hide/show menu via `InspectorMenuOpenContext`). It's correct but copy-prone — each future selection-dependent feature must remember to gate or it silently reintroduces a per-move O(N) recompute. A shared selection accessor that's inert unless a consumer opts in would make this structural.
- **Extract a `useCoalescedRafCallback` hook.** The rAF-coalescing logic is hand-rolled twice (`use-plot.ts` `scheduleSelectionRefresh`, `use-selected-rows.ts` `scheduleSyncRowSelectionToRdg`) with the same ref-guard + unmount-cancel shape.
- **De-duplicate the selectCases notification builder.** `selectCasesNotificationForDelta` repeats the result-assembly tail of `selectCasesNotification` (incl. the V2 "`undefined` not `[]`" convention). One could delegate to the other so that convention lives in one place.
- **Squares mount/trigger fragility.** The squares `<g>` is conditionally mounted while the `updateSquares` mstAutorun fires pre-commit — the fix is a targeted post-commit redraw effect. A sturdier long-term shape would be to always-mount the `<g>` (draw/clear instead of mount/unmount), removing the ref-timing race entirely.
- **Delta-hint multi-plot coverage.** `captureSelectionDelta` probes only `plotNum: 0`, so a two-y-attribute scatter falls back to a full restyle (safe, but loses the optimization). Checking all plot numbers would extend the fast path to multi-plot graphs.
- **`excludeTileId` parity.** The interactive notify path omits `notifyTileId`/`excludeTileId`. Harmless today (both paths broadcast), but a latent divergence if selection notifications ever start excluding the originating tile.
